xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,295 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+                         param_dims_per_shard[k].begin(), lower.begin(),
+                         std::minus<int64_t>());
+        }
+      } else {
+        fake_args_sliced[k].push_back(fake_args[k].Clone());
+      }
+    }
+    for (int k = 0; k < num_params; ++k) {
+      for (int i = 0; i < num_partitions; ++i) {
+        if (!param_sharded_dims[k].empty()) {
+          fake_ptrs[i].push_back(&fake_args_sliced[k][i]);
+        } else {
+          fake_ptrs[i].push_back(&fake_args_sliced[k][0]);
+        }
+      }
+    }
+
+    DeviceAssignment assn(/*replica_count=*/1,
+                          /*computation_count=*/num_partitions);
+    for (int64_t i = 0; i < num_partitions; ++i) {
+      assn(0, i) = i;
+    }
+    return HloTestBase::ExecuteReplicated(std::move(module), fake_ptrs,
+                                          num_partitions,
+                                          /*run_hlo_passes=*/true, &assn);
+  }
+
+  // Slice the unsharded reference results and compare to the sharded case.
+  void CompareShardedUnsharded(const std::string& hlo_text,
+                               int64_t num_partitions,
+                               std::vector<Literal>& ref_results,
+                               std::vector<Literal>& results,
+                               ErrorSpec& error_spec) {
+    HloModuleConfig config = GetModuleConfigForTest();
+    DebugOptions opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<VerifiedHloModule> module,
+                            ParseAndReturnVerifiedModule(hlo_text, config));
+    auto dimensions =
+        module->entry_computation()->root_instruction()->shape().dimensions();
+    std::vector<int64_t> root_dims(dimensions.begin(), dimensions.end());
+    std::vector<int64_t> root_dims_per_shard = root_dims;
+    std::vector<int64_t> root_sharded_dims;
+    {
+      HloSharding root_sharding =
+          module->entry_computation()->root_instruction()->sharding();
+      EvaluateShardedDims(root_dims_per_shard, root_sharded_dims,
+                          root_sharding);
+    }
+    if (!root_sharded_dims.empty()) {
+      std::vector<int64_t> lower(root_dims_per_shard.size(), 0);
+      std::vector<int64_t> upper(root_dims_per_shard.begin(),
+                                 root_dims_per_shard.end());
+      for (const Literal& result : results) {
+        Literal ref_results_slice = ref_results[0].Slice(lower, upper);
+        EXPECT_TRUE(
+            LiteralTestUtil::Near(ref_results_slice, result, error_spec));
+        for (int m = root_sharded_dims.size() - 1; m >= 0; --m) {
+          if (upper[root_sharded_dims[m]] < root_dims[root_sharded_dims[m]]) {
+            upper[root_sharded_dims[m]] +=
+                root_dims_per_shard[root_sharded_dims[m]];
+            break;
```

Comment:
Not the problem but this break is redundant. 

Also, this code looks somewhat similar to the above. Do you see a way to factor this out into a function and share them impl?